readthedocs/proxito/views/serve.py

```diff
@@ -95,17 +95,25 @@ def get(self,
         storage_path = final_project.get_storage_path(
             type_='html', version_slug=version_slug, include_file=False
         )
+
+        # URL without scheme and domain to perform an NGINX internal redirect
+        storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
+
+        # If ``filename`` is ``''`` it leaves a trailing slash
         path = os.path.join(storage_path, filename)
 
+        url = storage.url(path)  # this will remove the trailing slash
+        url = urlparse(url)._replace(scheme='', netloc='').geturl()
+
         # Handle our backend storage not supporting directory indexes,
         # so we need to append index.html when appropriate.
         if path[-1] == '/':
-            path += 'index.html'
+            url += '/index.html'
```

Comment:
We may check for `if filename == ''` here instead. Maybe is more readable.

Reply:
Presumably this logic is wrong now, if we have to re-add the slash. It was checking for URLs ending in `/`, where we need to then support a kind of  "directory index" -- so this isn't about missing filenames, but about any URL ending with a `/`.

Reply:
Oh, I see this is appending to a *different* path, that makes more sense now.

Reply:
Yes. It checks for `path` but adds it to `url`.

The problem is that we are mixing the usage of `os.path.join` and `storage.url`. Unfortunately, `storage.url` will always return a URL without the trailing slash when filename is `''`, but `os.path.join` won't. So, we use both for this reason here.

It's preferable to use `storage.url` because it will generate the proper URL using the storage backend system (and our mixins) --which will help us on corporate to generate the URL with the `AccessKey` to allow access to private docs.

Reply:
> Unfortunately, `storage.url` will always return a URL without the trailing slash when filename is `''`, but `os.path.join` won't. So, we use both for this reason here.

It seems we are forcing this at https://github.com/readthedocs/readthedocs.org/blob/1de81ca1f617bb4ea67b7cfd4a68985fdfb5fe51/readthedocs/builds/storage.py#L25-L36 for some reason

We could refactor this in another PR if we want.